Allow TextStats length distribution to be token-based and refactor for testability

Jauntbox opened the pull request:
**Related issues**
n/a

**Describe the proposed solution**
Tests did not catch that the token length distributions added to TextStats were actually entry length distributions. This PR refactors some of the functions in TextTokenizer, SmartTextVectorizer, and SmartTextMapVectorizer so that they are directly testable. It also adds more robust tests to check desired behavior of the TextStats object.

**Describe alternatives you've considered**
n/a

**Additional context**
n/a

tovbinm commented on core/src/main/scala/com/salesforce/op/stages/impl/feature/SmartTextMapVectorizer.scala:
```diff
@@ -203,6 +192,26 @@ class SmartTextMapVectorizer[T <: OPMap[String]]
 
+object SmartTextMapVectorizer extends CleanTextFun {
+
+  private[op] def computeTextMapStats[T <:  OPMap[String]](
```
please add docs for the function

tovbinm commented on core/src/main/scala/com/salesforce/op/stages/impl/feature/SmartTextVectorizer.scala:
```diff
@@ -164,13 +158,31 @@ class SmartTextVectorizer[T <: Text](uid: String = UID[SmartTextVectorizer[T]])(
     (all.collect { case (item, true) => item }, all.collect { case (item, false) => item })
   }
+
+  private[op] def computeTextStats[T <: Text : TypeTag](
```
same here

tovbinm commented on core/src/main/scala/com/salesforce/op/stages/impl/feature/SmartTextVectorizer.scala:
```diff
@@ -189,8 +201,8 @@ private[op] case class TextStats
     (acc, el) => acc + el._2 * (el._1 - lengthMean) * (el._1 - lengthMean)
-  )
-  val lengthStdDev: Double = math.sqrt(lengthVariance / lengthSize)
+  ) / lengthSize
```
can `lengthSize` be `0`?

Jauntbox replied:
Yep, and I think this is the behavior we want (these derived quantities being Double.NaN in this case).

The aggregations are already on the maps themselves, so these results being NaN will not propagate through - they are just descriptives. In the unlikely case where an entire column consists of stop words or something that doesn't tokenize (so lengthSize = 0), then we won't trigger removal either since Double.NaN < minLengthStdDev will always be false.

I can add a test where we check the values when lengthSize is 0 to make it more explicit.

tovbinm replied:
Ah, good point on NaN.

tovbinm commented on core/src/main/scala/com/salesforce/op/stages/impl/feature/TextTokenizer.scala:
```diff
@@ -162,30 +162,84 @@ object TextTokenizer {
+   * @param minTokenLength      minimum token length
+   * @return detected language and sentence tokens
+   */
+  def tokenize(
```
That's a bummer that you have to add so much boiler plate here. There are alternatives we can discuss.

Jauntbox replied:
Yes, I was not too happy with this either, but could not think of a clean alternative.

tovbinm replied:
Before this change everything had to be wrapped into `Text` to call the `tokenize`. Then for `String` value would have require additional wrapping into `Option`, which was safer but less efficient. 
I cant find where do you use other overloading for `tokenize` with `String` and `Option[String]`. Is it in this PR?

Jauntbox replied:
The `computeTextStats` and `computeTextMapStats` use them. I wanted to avoid having to wrap/unwrap values here since these functions are called per-entry.

tovbinm replied:
Makes sense then.

tovbinm commented on utils/src/main/scala/com/salesforce/op/test/TestCommon.scala:
```diff
@@ -130,4 +130,22 @@ trait TestCommon extends Matchers with Assertions {
+      if (expected != 0) {
+        math.abs((actual - expected) / expected) < tol shouldBe true
+      }
+      else math.abs(actual - expected) < tol shouldBe true
```
`math.abs(actual - expected) should be < tol`

Jauntbox replied:
Ahh, I tried shouldBe < tol (which did not work) - that's better

tovbinm commented on utils/src/main/scala/com/salesforce/op/test/TestCommon.scala:
```diff
@@ -130,4 +130,22 @@ trait TestCommon extends Matchers with Assertions {
+  def compareWithTol(actual: Double, expected: Double, tol: Double): Assertion =
+    withClue(s"Real number comparison failure! Actual: $actual, Expected: $expected \n") {
+      if (expected != 0) {
+        math.abs((actual - expected) / expected) < tol shouldBe true
```
`math.abs((actual - expected) / expected) should be < tol`

tovbinm commented on utils/src/main/scala/com/salesforce/op/test/TestCommon.scala:
```diff
@@ -130,4 +130,22 @@ trait TestCommon extends Matchers with Assertions {
+   * @return          Assertion to check approximate equality between actual and expected
+   */
+  def compareWithTol(actual: Double, expected: Double, tol: Double): Assertion =
+    withClue(s"Real number comparison failure! Actual: $actual, Expected: $expected \n") {
```
nice

tovbinm commented on core/src/main/scala/com/salesforce/op/stages/impl/feature/TextTokenizer.scala:
```diff
@@ -151,8 +151,8 @@ object TextTokenizer {
-  def tokenize(
-    text: Text,
+  def tokenizeString(
+    textString: String,
```
now this function can explode with NullPointerException is textString is null, while before it could have happened.

Jauntbox replied:
Hmmm, I'm not sure if it's possible for the `textString` argument to be `null` in practice though. When this function used for tokenizing the map entries, a value that was originally null there will just not show up as an entry in the map. When it's used for tokenizing a normal Text entry, then we should have already safely converted any nulls or missing elements into an `Option[String]`, right?

The actual tokenize call during vectorization is still `tokenize(v.toText)` where `v` is the value in a text map. I'd actually argue that that should be changed to `tokenizeText(v)` to save time converting it to Text and back again.

I agree it's technically less safe, but I don't think it's necessary to have null checking at this point in the flow. We should make sure the data gets created in a safe way, which I think we already do. Are there some specific edge cases I'm missing?

tovbinm replied:
I think the simplest is to add a null check `tokenizeString` and return

Jauntbox replied:
Now that I think about it more, I'm pretty sure the old tokenize function would also give a NPE if you fed it a sneaky null value. The `SomeValue.unapply` function explicitly calls `v.isEmpty` which would also fail if v was `null`.

I put back the old tokenize function as `oldTokenize` and tried
```val sneakyStringOpt: Option[String] = null
val myText = Text(sneakyStringOpt)
val res = TextTokenizer.oldTokenize(myText)```
which did indeed throw a NPE.

We have tests all over the place (eg. our vectorizer tests and FeatureTypeSparkConverterTest) that make sure we can handle `null` values in dataframes and safely convert them into our types. I'm not aware of any explicit null checks in our functions elsewhere, so it just feels weird to put one here.

@leahmcguire any opinions on this?

tovbinm replied:
SomeValue.unapply operates on value which is Option[String]. Null check is done during the construction of Text.

tovbinm replied:
Your example you provided is not currently possible and also not a fair one :)

leahmcguire replied:
@Jauntbox is this only called from the Option[String] version below? if so make it private and it is fine.

leahmcguire replied:
In fact make them both private please

TuanNguyen27 commented on core/src/main/scala/com/salesforce/op/filters/FeatureDistribution.scala:
```diff
@@ -282,7 +283,15 @@ object FeatureDistribution {
+      case Right(doubleSeq) => doubleSeq.map(_.toString)
+    }
+    stringVals.foldLeft(TextStats.empty)((acc, el) => acc + SmartTextVectorizer.computeTextStats(
+      Option(el), shouldCleanText = false, maxCardinality = RawFeatureFilter.MaxCardinality)
```
Should `shouldCleanText = true` instead ?

Jauntbox replied:
I can change it, but I don't think it matters much here. These values aren't used in SmartTextVectorizer. They're the ones that show up in the ModelInsights.

TuanNguyen27 commented on core/src/main/scala/com/salesforce/op/stages/impl/feature/SmartTextVectorizer.scala:
```diff
@@ -164,13 +158,40 @@ class SmartTextVectorizer[T <: Text](uid: String = UID[SmartTextVectorizer[T]])(
+      .foldLeft(Map.empty[Int, Long])(
+        (acc, el) => TextStats.additionHelper(acc, Map(el.length -> 1L), maxCardinality)
+      )
+    val (valueCounts, lengthCounts) = text match {
```
when we reach `RawFeatureFilter.MaxCardinality` for `valueCounts`, will `lengthCounts` also stop accumulating ?

TuanNguyen27 replied:
I think in my earlier attempt to track token length distribution i also ran into this issue...

TuanNguyen27 replied:
nvm, this is taken care of by `val newLengthCounts = additionHelper(l.lengthCounts, r.lengthCounts, maxCardinality)` , pls disregard this comment :D

TuanNguyen27 commented on core/src/main/scala/com/salesforce/op/stages/impl/feature/TextTokenizer.scala:
```diff
@@ -162,30 +162,84 @@ object TextTokenizer {
+      .getOrElse(Seq(lowerTxt))
+      .map { sentence =>
+        val tokens = analyzer.analyze(sentence, language)
+        tokens.filter(_.length >= minTokenLength).toTextList
```
Why are we only keeping tokens with length > `minTokenLength ` ?

Jauntbox replied:
This was existing behavior. It's a configurable parameter (defaulting to 1), so is not required.

Jauntbox commented:
Just a heads up on a few more commits - adding a toggle for tokenization in text lengths. Will cause a problems with Chinese/Korean text based on our current tokenizers.

Jauntbox commented:
Ok - ready for a final look. Sorry for the last-minute refactoring, but realized we needed this toggle exposed for experiments.

Final refactoring:
- Removed tokenizeStringOpt method since we can get by without it (should be more readable too)
- Added toggle to SmartTextVectorizer and SmartTextMapVectorizer to enable/disable tokenization when calculating length distributions in TextStats
- Added tests to check that this toggle does what we want
- Added logging of derived quantities and vectorization method used in SmartTextVectorizer and SmartTextMapVectorizer

TuanNguyen27 commented on core/src/main/scala/com/salesforce/op/filters/FeatureDistribution.scala:
```diff
@@ -268,21 +271,36 @@ object FeatureDistribution {
+    )
   }
 
   private def countStringValues[T](seq: Seq[T]): Map[String, Long] = {
```
not relevant to this PR but i think `countStringValues` is no longer used.

Jauntbox replied:
ok, I can remove it then

TuanNguyen27 commented on core/src/main/scala/com/salesforce/op/filters/PreparedFeatures.scala:
```diff
@@ -169,7 +169,8 @@ private[filters] object PreparedFeatures {
       case SomeValue(v: SparseVector) => Map((name, None) -> Right(v.indices.map(_.toDouble).toSeq))
       case ft@SomeValue(_) => ft match {
-        case v: Text => Map((name, None) -> Left(v.value.toSeq.flatMap(tokenize)))
+        // case v: Text => Map((name, None) -> Left(v.value.toSeq.flatMap(tokenize)))
```
We are no longer tokenzing text during data prep?

Jauntbox replied:
Whoops, that was for testing - forgot to take it out.

TuanNguyen27 commented:
lgtm !

leahmcguire commented on core/src/main/scala/com/salesforce/op/dsl/RichMapFeature.scala:
```diff
@@ -322,6 +328,8 @@ trait RichMapFeature {
         .setHashSpaceStrategy(hashSpaceStrategy)
         .setHashAlgorithm(hashAlgorithm)
         .setBinaryFreq(binaryFreq)
+        .setTokenizeForLengths(tokenizeForLengths)
```
can we make this an enum rather than a boolean? then we have room to expend in the future
